File: packages/sdk/src/call.ts
import {
  Room,
  RoomEvent,
  LocalParticipant,
  RemoteParticipant,
  ConnectionQuality as LKQuality,
  type Participant as LKParticipant,
  Track,
} from 'livekit-client'
import { EventEmitter } from './events.js'
import type {
  CallOptions,
  CallEventMap,
  ConnectionState,
  Participant,
  Message,
  TranscriptSegment,
  DeviceList,
  Layout,
  ConnectionQuality,
} from './types.js'

function mapQuality(q: LKQuality): ConnectionQuality {
  switch (q) {
    case LKQuality.Excellent: return 'excellent'
    case LKQuality.Good: return 'good'
    case LKQuality.Poor: return 'poor'
    case LKQuality.Lost: return 'lost'
    default: return 'unknown'
  }
}

function mapParticipant(p: LKParticipant, isLocal = false): Participant {
  let meta: Record<string, unknown> = {}
  try { meta = p.metadata ? JSON.parse(p.metadata) : {} } catch { /* ignore */ }

  const videoTrack = p.getTrackPublication(Track.Source.Camera)?.track?.mediaStreamTrack ?? null
  const audioTrack = p.getTrackPublication(Track.Source.Microphone)?.track?.mediaStreamTrack ?? null
  const screenSharePub = p.getTrackPublication(Track.Source.ScreenShare)
  const screenShareTrack = screenSharePub?.track?.mediaStreamTrack ?? null

  return {
    id: p.identity,
    name: p.name ?? p.identity,
    role: (meta['role'] as Participant['role']) ?? 'participant',
    isMuted: !p.isMicrophoneEnabled,
    isCameraOff: !p.isCameraEnabled,
    isSpeaking: p.isSpeaking,
    connectionQuality: mapQuality(p.connectionQuality),
    videoTrack,
    audioTrack,
    screenShareTrack,
    isScreenSharing: screenSharePub?.isSubscribed !== false && screenShareTrack !== null,
    isLocal,
    metadata: meta,
  }
}

let _msgCounter = 0

export class Call extends EventEmitter<CallEventMap> {
  private room: Room
  private options: CallOptions
  private _connectionState: ConnectionState = 'idle'
  private _participants = new Map<string, Participant>()
  private _activeSpeakers: Participant[] = []
  private _messages: Message[] = []
  private _devices: DeviceList = { audioinput: [], audiooutput: [], videoinput: [] }
  private _layout: Layout = 'grid'
  private _pinnedParticipant: string | null = null
  private _tokenExpiresAt: Date = new Date(0)

  constructor(options: CallOptions) {
    super()
    this.options = options
    this.room = new Room()
    this._setupRoomListeners()
  }

  get connectionState(): ConnectionState { return this._connectionState }
  get participants(): Map<string, Participant> { return this._participants }
  get localParticipant(): Participant | null {
    const lp = this.room.localParticipant
    return lp ? mapParticipant(lp, true) : null
  }
  get activeSpeakers(): Participant[] { return this._activeSpeakers }
  get messages(): Message[] { return this._messages }
  get devices(): DeviceList { return this._devices }
  get layout(): Layout { return this._layout }
  get pinnedParticipant(): string | null { return this._pinnedParticipant }
  get tokenExpiresAt(): Date { return this._tokenExpiresAt }
  get livekitUrl(): string { return this.options.url }

  private setState(state: ConnectionState) {
    this._connectionState = state
    this.emit('connectionStateChanged', state)
  }

  private _syncParticipants(emitUpdates = false) {
    this._participants.clear()
    for (const [, p] of this.room.remoteParticipants) {
      const mapped = mapParticipant(p)
      this._participants.set(p.identity, mapped)
      if (emitUpdates) this.emit('participantUpdated', mapped)
    }
    if (emitUpdates) {
      const local = this.localParticipant
      if (local) this.emit('participantUpdated', local)
    }
  }

  private _setupRoomListeners() {
    const { room } = this

    room.on(RoomEvent.Connected, () => {
      this.setState('connected')
      this._syncParticipants()
    })

    room.on(RoomEvent.Disconnected, (reason) => {
      this.setState('disconnected')
      this.emit('disconnected', reason?.toString())
    })

    room.on(RoomEvent.Reconnecting, () => {
      this.setState('reconnecting')
      this.emit('reconnecting', 1)
    })

    room.on(RoomEvent.Reconnected, () => {
      this.setState('connected')
      this.emit('reconnected')
    })

    room.on(RoomEvent.ParticipantConnected, (p: RemoteParticipant) => {
      const participant = mapParticipant(p)
      this._participants.set(p.identity, participant)
      this.emit('participantJoined', participant)
    })

    room.on(RoomEvent.ParticipantDisconnected, (p: RemoteParticipant) => {
      const participant = this._participants.get(p.identity) ?? mapParticipant(p)
      this._participants.delete(p.identity)
      this.emit('participantLeft', participant)
    })

    room.on(RoomEvent.ActiveSpeakersChanged, (speakers: LKParticipant[]) => {
      this._activeSpeakers = speakers.map((s) => mapParticipant(s, s instanceof LocalParticipant))
      this.emit('activeSpeakerChanged', this._activeSpeakers)
      this._syncParticipants(true)
    })

    room.on(RoomEvent.TrackPublished, (pub, participant) => {
      this._syncParticipants(true)
      if (pub.source === Track.Source.ScreenShare) {
        const mapped = mapParticipant(participant)
        this.emit('screenShareStarted', mapped)
      }
    })

    room.on(RoomEvent.TrackUnpublished, (pub, participant) => {
      this._syncParticipants(true)
      if (pub.source === Track.Source.ScreenShare) {
        this.emit('screenShareStopped', participant.identity)
      }
    })

    room.on(RoomEvent.TrackSubscribed, (track, pub, participant) => {
      this._syncParticipants(true)
      if (pub.source === Track.Source.ScreenShare) {
        const mapped = mapParticipant(participant)
        this.emit('screenShareStarted', mapped)
      }
    })

    room.on(RoomEvent.TrackUnsubscribed, (track, pub, participant) => {
      this._syncParticipants(true)
      if (pub.source === Track.Source.ScreenShare) {
        this.emit('screenShareStopped', participant.identity)
      }
    })

    room.on(RoomEvent.TrackMuted, () => this._syncParticipants(true))
    room.on(RoomEvent.TrackUnmuted, () => this._syncParticipants(true))
    room.on(RoomEvent.ParticipantMetadataChanged, () => this._syncParticipants(true))
    room.on(RoomEvent.ConnectionQualityChanged, () => this._syncParticipants(true))

    room.on(RoomEvent.LocalTrackPublished, (pub) => {
      const local = this.localParticipant
      if (local) this.emit('participantUpdated', local)
      if (pub.source === Track.Source.ScreenShare) {
        if (local) this.emit('screenShareStarted', local)
      }
    })

    room.on(RoomEvent.LocalTrackUnpublished, (pub) => {
      const local = this.localParticipant
      if (local) this.emit('participantUpdated', local)
      if (pub.source === Track.Source.ScreenShare) {
        this.emit('screenShareStopped', this.room.localParticipant.identity)
      }
    })

    room.on(RoomEvent.RecordingStatusChanged, (recording: boolean) => {
      if (recording) this.emit('recordingStarted')
      else this.emit('recordingStopped')
    })

    room.on(RoomEvent.AudioPlaybackStatusChanged, () => {
      if (!room.canPlaybackAudio) this.emit('audioPlaybackBlocked')
    })

    room.on(RoomEvent.DataReceived, (payload: Uint8Array, participant?: RemoteParticipant) => {
      try {
        const text = new TextDecoder().decode(payload)
        const data = JSON.parse(text) as { type: string; text?: string; emoji?: string; id?: string; speakerId?: string; speaker?: string; startMs?: number }

        if (data.type === 'chat' && data.text) {
          const msg: Message = {
            id: data.id ?? String(++_msgCounter),
            from: participant?.identity ?? 'unknown',
            fromName: participant?.name ?? participant?.identity ?? 'unknown',
            text: data.text,
            timestamp: new Date(),
            to: null,
          }
          this._messages = [...this._messages.slice(-499), msg]
          this.emit('messageReceived', msg)
        } else if (data.type === 'reaction' && data.emoji) {
          this.emit('reactionReceived', participant?.identity ?? 'unknown', data.emoji)
        } else if (data.type === 'speaking') {
          const speakerId = data.speakerId ?? participant?.identity ?? 'unknown'
          const speakerName = data.speaker ?? participant?.name ?? participant?.identity ?? 'unknown'
          this.emit('speakingStarted', speakerId, speakerName)
        } else if (data.type === 'transcript' && data.text) {
          const speakerId = data.speakerId ?? participant?.identity ?? 'unknown'
          this.emit('speakingStopped', speakerId)
          const segment: TranscriptSegment = {
            text: data.text as string,
            speaker: data.speaker ?? participant?.name ?? participant?.identity ?? 'unknown',
            speakerId,
            timestamp: new Date(),
            ...(data.startMs !== undefined && { startMs: data.startMs }),
          }
          this.emit('transcriptReceived', segment)
        }
      } catch { /* ignore malformed data */ }
    })
  }

  async connect(): Promise<void> {
    this.setState('connecting')
    let token = this.options.token

    if (this.options.tokenRefresher) {
      const expires = this._parseTokenExpiry(token)
      this._tokenExpiresAt = expires
      if (Date.now() >= expires.getTime() - 60_000) {
        token = await this.options.tokenRefresher()
        this._tokenExpiresAt = this._parseTokenExpiry(token)
      }
    }

    await this.room.connect(this.options.url, token)
    await this._enumerateDevices()
  }

  async disconnect(): Promise<void> {
    await this.room.disconnect()
    this.setState('disconnected')
  }

  async toggleMic(): Promise<void> {
    await this.room.localParticipant.setMicrophoneEnabled(
      !this.room.localParticipant.isMicrophoneEnabled
    )
  }

  async setMicEnabled(enabled: boolean): Promise<void> {
    await this.room.localParticipant.setMicrophoneEnabled(enabled)
  }

  async toggleCamera(): Promise<void> {
    await this.room.localParticipant.setCameraEnabled(
      !this.room.localParticipant.isCameraEnabled
    )
  }

  async setCameraEnabled(enabled: boolean): Promise<void> {
    await this.room.localParticipant.setCameraEnabled(enabled)
  }

  async startScreenShare(): Promise<void> {
    await this.room.localParticipant.setScreenShareEnabled(true)
  }

  async stopScreenShare(): Promise<void> {
    await this.room.localParticipant.setScreenShareEnabled(false)
  }

  isScreenSharing(): boolean {
    return !!this.room.localParticipant?.getTrackPublication(Track.Source.ScreenShare)
  }

  async switchDevice(kind: 'audioinput' | 'audiooutput' | 'videoinput', deviceId: string): Promise<void> {
    await this.room.switchActiveDevice(kind, deviceId)
  }

  async sendMessage(text: string, options?: { to?: string[] }): Promise<void> {
    const data = JSON.stringify({ type: 'chat', text, id: String(++_msgCounter) })
    const payload = new TextEncoder().encode(data)
    if (options?.to?.length) {
      await this.room.localParticipant.publishData(payload, {
        reliable: true,
        destinationIdentities: options.to,
      })
    } else {
      await this.room.localParticipant.publishData(payload, { reliable: true })
    }
  }

  async sendReaction(emoji: string): Promise<void> {
    const data = JSON.stringify({ type: 'reaction', emoji })
    const payload = new TextEncoder().encode(data)
    await this.room.localParticipant.publishData(payload, { reliable: true })
  }

  setLayout(layout: Layout): void {
    this._layout = layout
  }

  pin(participantId: string | null): void {
    this._pinnedParticipant = participantId
  }

  private _requireApi(): { apiUrl: string; roomName: string } {
    const { apiUrl, roomName } = this.options
    if (!apiUrl) throw new Error('CallOptions.apiUrl is required for recording/transcription control')
    if (!roomName) throw new Error('CallOptions.roomName is required for recording/transcription control')
    return { apiUrl, roomName }
  }

  private async _apiPost(path: string): Promise<void> {
    const { apiUrl, roomName } = this._requireApi()
    const url = `${apiUrl.replace(/\/$/, '')}/rooms/${encodeURIComponent(roomName)}${path}`
    const res = await fetch(url, { method: 'POST', headers: { 'Content-Type': 'application/json' } })
    if (!res.ok) throw new Error(`API request failed: ${res.status} ${res.statusText}`)
  }

  async startRecording(): Promise<void> {
    await this._apiPost('/recording/start')
  }

  async stopRecording(): Promise<void> {
    await this._apiPost('/recording/stop')
  }

  async startTranscription(): Promise<void> {
    await this._apiPost('/transcription/start')
  }

  async stopTranscription(): Promise<void> {
    await this._apiPost('/transcription/stop')
  }

  private _parseTokenExpiry(token: string): Date {
    try {
      const payload = JSON.parse(atob(token.split('.')[1]!)) as { exp?: number }
      return payload.exp ? new Date(payload.exp * 1000) : new Date(Date.now() + 6 * 3600 * 1000)
    } catch {
      return new Date(Date.now() + 6 * 3600 * 1000)
    }
  }

  private async _enumerateDevices(): Promise<void> {
    try {
      const devices = await navigator.mediaDevices.enumerateDevices()
      this._devices = {
        audioinput: devices.filter((d) => d.kind === 'audioinput'),
        audiooutput: devices.filter((d) => d.kind === 'audiooutput'),
        videoinput: devices.filter((d) => d.kind === 'videoinput'),
      }
      this.emit('devicesChanged', this._devices)
    } catch { /* permissions not granted yet */ }
  }
}

File: mobile/android/sdk/src/main/kotlin/com/rtcstack/sdk/Call.kt
package com.rtcstack.sdk

import android.content.Context
import android.content.Intent
import android.util.Base64
import io.livekit.android.LiveKit
import io.livekit.android.events.RoomEvent
import io.livekit.android.events.collect
import io.livekit.android.room.Room
import io.livekit.android.room.participant.Participant as LKParticipant
import io.livekit.android.room.participant.ConnectionQuality as LKQuality
import io.livekit.android.room.track.AudioTrack
import io.livekit.android.room.track.Track
import io.livekit.android.room.track.VideoTrack
import io.livekit.android.room.track.screencapture.ScreenCaptureParams
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import org.json.JSONObject
import java.util.concurrent.atomic.AtomicLong

/**
 * The primary RTCstack call object — a thin, idiomatic Kotlin wrapper over a LiveKit [Room],
 * mirroring the shipped `@rtcstack/sdk` `Call` class (`packages/sdk/src/call.ts`).
 *
 * Reactive surface:
 *  - [events]: one-shot [CallEvent]s (joined/left/message/etc.).
 *  - [connectionState], [participants], [localParticipant], [activeSpeakers], [messages],
 *    [layout], [pinnedParticipant]: snapshot StateFlows the UI observes.
 *
 * Construct via [RTCstack.createCall]. Does NOT connect until [connect] is called.
 */
public class Call internal constructor(
    private val appContext: Context,
    private val options: CallOptions,
) {
    private val room: Room = LiveKit.create(appContext)
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private var roomEventsJob: Job? = null
    private val msgCounter = AtomicLong(0)

    private val _events = MutableSharedFlow<CallEvent>(extraBufferCapacity = 64)
    public val events: SharedFlow<CallEvent> = _events.asSharedFlow()

    private val _connectionState = MutableStateFlow(ConnectionState.IDLE)
    public val connectionState: StateFlow<ConnectionState> = _connectionState.asStateFlow()

    private val _participants = MutableStateFlow<Map<String, Participant>>(emptyMap())
    public val participants: StateFlow<Map<String, Participant>> = _participants.asStateFlow()

    private val _localParticipant = MutableStateFlow<Participant?>(null)
    public val localParticipant: StateFlow<Participant?> = _localParticipant.asStateFlow()

    private val _activeSpeakers = MutableStateFlow<List<Participant>>(emptyList())
    public val activeSpeakers: StateFlow<List<Participant>> = _activeSpeakers.asStateFlow()

    private val _messages = MutableStateFlow<List<Message>>(emptyList())
    public val messages: StateFlow<List<Message>> = _messages.asStateFlow()

    private val _layout = MutableStateFlow(Layout.GRID)
    public val layout: StateFlow<Layout> = _layout.asStateFlow()

    private val _pinned = MutableStateFlow<String?>(null)
    public val pinnedParticipant: StateFlow<String?> = _pinned.asStateFlow()

    /** Token expiry (epoch millis), decoded from the JWT at connect time. */
    public var tokenExpiresAt: Long = 0L
        private set

    public val livekitUrl: String get() = options.url

    /** Underlying LiveKit room — exposed for the UI kit (components-android) to render tracks. */
    public val lkRoom: Room get() = room

    // ─── Connection ──────────────────────────────────────────────────────────

    public suspend fun connect() {
        setState(ConnectionState.CONNECTING)
        options.onAnalyticsEvent?.invoke("connect_start", mapOf("url" to options.url))

        var token = options.token
        tokenExpiresAt = parseTokenExpiry(token)

        val refresher = options.tokenRefresher
        if (refresher != null && System.currentTimeMillis() >= tokenExpiresAt - 60_000) {
            token = refresher()
            tokenExpiresAt = parseTokenExpiry(token)
        }

        startObservingRoom()
        try {
            room.connect(options.url, token)
        } catch (e: Exception) {
            setState(ConnectionState.DISCONNECTED)
            val code = classifyConnectError(e)
            _events.tryEmit(CallEvent.Error(code, e.message ?: "connect failed"))
            throw CallException(code, e.message ?: "connect failed", e)
        }
    }

    public suspend fun disconnect() {
        room.disconnect()
        setState(ConnectionState.DISCONNECTED)
        roomEventsJob?.cancel()
    }

    /** Releases the room and coroutine scope. Call when the call object is no longer needed. */
    public fun release() {
        roomEventsJob?.cancel()
        room.release()
    }

    // ─── Media controls ──────────────────────────────────────────────────────

    public suspend fun toggleMic() {
        room.localParticipant.setMicrophoneEnabled(!room.localParticipant.isMicrophoneEnabled)
    }

    public suspend fun setMicEnabled(enabled: Boolean) {
        room.localParticipant.setMicrophoneEnabled(enabled)
    }

    public suspend fun toggleCamera() {
        room.localParticipant.setCameraEnabled(!room.localParticipant.isCameraEnabled)
    }

    public suspend fun setCameraEnabled(enabled: Boolean) {
        room.localParticipant.setCameraEnabled(enabled)
    }

    /**
     * Start screen share. Requires the MediaProjection consent Intent from
     * [com.rtcstack.sdk.screenshare.ScreenShareLauncher] AND a running
     * [com.rtcstack.sdk.call.RTCstackCallService] foreground service of type mediaProjection.
     */
    public suspend fun startScreenShare(mediaProjectionPermissionResultData: Intent) {
        room.localParticipant.setScreenShareEnabled(
            true,
            ScreenCaptureParams(mediaProjectionPermissionResultData),
        )
    }

    public suspend fun stopScreenShare() {
        room.localParticipant.setScreenShareEnabled(false)
    }

    public fun isScreenSharing(): Boolean =
        room.localParticipant.getTrackPublication(Track.Source.SCREEN_SHARE)?.track != null

    // ─── Messaging ───────────────────────────────────────────────────────────

    /**
     * Send a chat message. NOTE: matches the shipped web SDK — LiveKit does NOT loop data
     * back to the sender, so neither [messages] nor a [CallEvent.MessageReceived] fires
     * locally. The UI layer is responsible for echoing the user's own outgoing messages.
     */
    public suspend fun sendMessage(text: String, to: List<String>? = null) {
        val id = msgCounter.incrementAndGet().toString()
        publish(WireFormat.encodeChat(text, id), to)
    }

    /** Send a reaction. Like [sendMessage], no local echo — the UI shows the sender's own reaction. */
    public suspend fun sendReaction(emoji: String) {
        publish(WireFormat.encodeReaction(emoji), null)
    }

    private suspend fun publish(payload: ByteArray, to: List<String>?) {
        val identities = to?.map { LKParticipant.Identity(it) }
        // NOTE(verify-on-mac): confirm publishData signature for the pinned livekit-android version.
        room.localParticipant.publishData(data = payload, identities = identities)
    }

    // ─── Layout state ────────────────────────────────────────────────────────

    public fun setLayout(layout: Layout) { _layout.value = layout }
    public fun pin(participantId: String?) { _pinned.value = participantId }

    // ─── Internals ───────────────────────────────────────────────────────────

    private fun setState(state: ConnectionState) {
        _connectionState.value = state
        _events.tryEmit(CallEvent.ConnectionStateChanged(state))
    }

    private fun startObservingRoom() {
        roomEventsJob?.cancel()
        roomEventsJob = scope.launch {
            room.events.collect { event -> handleRoomEvent(event) }
        }
    }

    private fun handleRoomEvent(event: RoomEvent) {
        when (event) {
            is RoomEvent.Connected -> {
                setState(ConnectionState.CONNECTED)
                options.onAnalyticsEvent?.invoke("connect_success", emptyMap())
                syncParticipants()
                syncLocal()
            }
            is RoomEvent.Disconnected -> {
                setState(ConnectionState.DISCONNECTED)
                _events.tryEmit(CallEvent.Disconnected(mapDisconnect(event)))
            }
            is RoomEvent.Reconnecting -> {
                setState(ConnectionState.RECONNECTING)
                _events.tryEmit(CallEvent.Reconnecting(1))
            }
            is RoomEvent.Reconnected -> {
                setState(ConnectionState.CONNECTED)
                _events.tryEmit(CallEvent.Reconnected)
            }
            is RoomEvent.ParticipantConnected -> {
                val p = mapParticipant(event.participant)
                _participants.value = _participants.value + (p.id to p)
                _events.tryEmit(CallEvent.ParticipantJoined(p))
            }
            is RoomEvent.ParticipantDisconnected -> {
                val id = event.participant.identity?.value ?: return
                val existing = _participants.value[id] ?: mapParticipant(event.participant)
                _participants.value = _participants.value - id
                _events.tryEmit(CallEvent.ParticipantLeft(existing))
            }
            is RoomEvent.ActiveSpeakersChanged -> {
                val speakers = event.speakers.map { mapParticipant(it) }
                _activeSpeakers.value = speakers
                _events.tryEmit(CallEvent.ActiveSpeakerChanged(speakers))
                syncParticipants(emitUpdates = true)
            }
            is RoomEvent.TrackPublished -> {
                syncParticipants(emitUpdates = true)
                if (event.publication.source == Track.Source.SCREEN_SHARE) {
                    _events.tryEmit(CallEvent.ScreenShareStarted(mapParticipant(event.participant)))
                }
            }
            is RoomEvent.TrackUnpublished -> {
                syncParticipants(emitUpdates = true)
                if (event.publication.source == Track.Source.SCREEN_SHARE) {
                    _events.tryEmit(CallEvent.ScreenShareStopped(event.participant.identity?.value ?: ""))
                }
            }
            is RoomEvent.TrackSubscribed -> syncParticipants(emitUpdates = true)
            is RoomEvent.TrackUnsubscribed -> syncParticipants(emitUpdates = true)
            is RoomEvent.TrackMuted -> syncParticipants(emitUpdates = true)
            is RoomEvent.TrackUnmuted -> syncParticipants(emitUpdates = true)
            is RoomEvent.ParticipantMetadataChanged -> syncParticipants(emitUpdates = true)
            is RoomEvent.ConnectionQualityChanged -> syncParticipants(emitUpdates = true)
            is RoomEvent.RecordingStatusChanged -> {
                // 2.18.2's RecordingStatusChanged exposes no property accessor; read off the Room.
                if (room.isRecording) _events.tryEmit(CallEvent.RecordingStarted)
                else _events.tryEmit(CallEvent.RecordingStopped)
            }
            is RoomEvent.DataReceived -> handleData(event)
            else -> Unit // forward-compat: ignore LiveKit events we don't surface
        }
    }

    private fun handleData(event: RoomEvent.DataReceived) {
        val sender = event.participant
        when (val inbound = WireFormat.decode(event.data)) {
            is WireFormat.Inbound.Chat -> {
                val msg = Message(
                    id = inbound.id ?: msgCounter.incrementAndGet().toString(),
                    from = sender?.identity?.value ?: "unknown",
                    fromName = sender?.name ?: sender?.identity?.value ?: "unknown",
                    text = inbound.text,
                    timestamp = System.currentTimeMillis(),
                    to = null,
                )
                _messages.value = (_messages.value + msg).takeLast(500)
                _events.tryEmit(CallEvent.MessageReceived(msg))
            }
            is WireFormat.Inbound.Reaction ->
                _events.tryEmit(CallEvent.ReactionReceived(sender?.identity?.value ?: "unknown", inbound.emoji))
            is WireFormat.Inbound.Speaking -> {
                val sid = inbound.speakerId ?: sender?.identity?.value ?: "unknown"
                val sname = inbound.speaker ?: sender?.name ?: sid
                _events.tryEmit(CallEvent.SpeakingStarted(sid, sname))
            }
            is WireFormat.Inbound.Transcript -> {
                val sid = inbound.speakerId ?: sender?.identity?.value ?: "unknown"
                _events.tryEmit(CallEvent.SpeakingStopped(sid))
                _events.tryEmit(
                    CallEvent.TranscriptReceived(
                        TranscriptSegment(
                            text = inbound.text,
                            speaker = inbound.speaker ?: sender?.name ?: sid,
                            speakerId = sid,
                            timestamp = System.currentTimeMillis(),
                            startMs = inbound.startMs,
                        ),
                    ),
                )
            }
            WireFormat.Inbound.Ignored, null -> Unit
        }
    }

    private fun syncParticipants(emitUpdates: Boolean = false) {
        val mapped = room.remoteParticipants.values.associate {
            val p = mapParticipant(it)
            p.id to p
        }
        _participants.value = mapped
        if (emitUpdates) {
            mapped.values.forEach { _events.tryEmit(CallEvent.ParticipantUpdated(it)) }
            syncLocal()
        }
    }

    private fun syncLocal() {
        val p = mapParticipant(room.localParticipant, isLocal = true)
        _localParticipant.value = p
        _events.tryEmit(CallEvent.ParticipantUpdated(p))
    }

    private fun mapParticipant(p: LKParticipant, isLocal: Boolean = false): Participant {
        val camera = p.getTrackPublication(Track.Source.CAMERA)?.track as? VideoTrack
        val mic = p.getTrackPublication(Track.Source.MICROPHONE)?.track as? AudioTrack
        val screen = p.getTrackPublication(Track.Source.SCREEN_SHARE)?.track as? VideoTrack
        val meta = parseMetadata(p.metadata)
        return Participant(
            id = p.identity?.value ?: "",
            name = p.name ?: p.identity?.value ?: "",
            role = ParticipantRole.fromString(meta["role"] as? String),
            isMuted = !p.isMicrophoneEnabled,
            isCameraOff = !p.isCameraEnabled,
            isSpeaking = p.isSpeaking,
            connectionQuality = mapQuality(p.connectionQuality),
            videoTrack = camera,
            audioTrack = mic,
            screenShareTrack = screen,
            isScreenSharing = screen != null,
            isLocal = isLocal,
            metadata = meta,
        )
    }

    private fun mapQuality(q: LKQuality): ConnectionQuality = when (q) {
        LKQuality.EXCELLENT -> ConnectionQuality.EXCELLENT
        LKQuality.GOOD -> ConnectionQuality.GOOD
        LKQuality.POOR -> ConnectionQuality.POOR
        LKQuality.LOST -> ConnectionQuality.LOST
        else -> ConnectionQuality.UNKNOWN
    }

    private fun mapDisconnect(event: RoomEvent.Disconnected): DisconnectReason {
        // NOTE(verify-on-mac): map event.reason (livekit DisconnectReason) precisely.
        return DisconnectReason.UNKNOWN
    }

    private fun classifyConnectError(e: Exception): CallErrorCode {
        val m = e.message?.lowercase().orEmpty()
        return when {
            "auth" in m || "401" in m || "token" in m -> CallErrorCode.AUTH_FAILED
            "not found" in m || "404" in m -> CallErrorCode.ROOM_NOT_FOUND
            else -> CallErrorCode.NETWORK_ERROR
        }
    }

    private fun parseTokenExpiry(token: String): Long {
        return try {
            val parts = token.split(".")
            val payload = String(Base64.decode(parts[1], Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP))
            val exp = JSONObject(payload).optLong("exp", 0L)
            if (exp > 0) exp * 1000 else System.currentTimeMillis() + 6 * 3600_000L
        } catch (_: Exception) {
            System.currentTimeMillis() + 6 * 3600_000L
        }
    }
}
